File: Go/day04.go
package main

import (
	"fmt"
	"os"
	"reflect"
	"strconv"
	"strings"
)

type cell struct {
	number int
	draw   bool
}

func remove(slice [][][]cell, s [][]cell) [][][]cell {
	for i := 0; i < len(slice); i++ {
		if reflect.DeepEqual(slice[i], s) {
			return append(slice[:i], slice[i+1:]...)
		}
	}
	return slice
}

func read_input() ([]int, [][][]cell) {
	input, _ := os.ReadFile("inputs/day04.txt")
	lines := strings.Split(string(input), "\n")
	nums := strings.Split(lines[0], ",")
	var new_nums []int
	for _, num := range nums {
		new_value, _ := strconv.Atoi(num)
		new_nums = append(new_nums, new_value)
	}
	tables := strings.Split(string(input), "\n\n")
	tables = tables[1:]
	var new_tables [][][]cell
	for _, table := range tables {
		table_slice := strings.Split(string(table), "\n")
		var new_table [][]cell
		for _, row := range table_slice {
			row = strings.Replace(row, "  ", " ", -1)
			row_slice := strings.Split(row, " ")
			if row_slice[0] == "" {
				row_slice = row_slice[1:]
			}
			var new_row []cell
			for _, column := range row_slice {
				var new_column cell
				new_column.draw = false
				new_column.number, _ = strconv.Atoi(column)
				new_row = append(new_row, new_column)
			}
			new_table = append(new_table, new_row)
		}
		new_tables = append(new_tables, new_table)
	}

	return new_nums, new_tables
}

func day04_I(tables [][][]cell, nums []int) {
	for _, num := range nums {
		for i, table := range tables {
			for j, row := range table {
				for k, column := range row {
					if column.number == num {
						tables[i][j][k].draw = true
					}
				}
			}
		}

		for _, table := range tables {
			for i := 0; i < 5; i++ {
				if (table[i][0].draw == true && table[i][1].draw == true && table[i][2].draw == true && table[i][3].draw == true && table[i][4].draw == true) || (table[0][i].draw == true && table[1][i].draw == true && table[2][i].draw == true && table[3][i].draw == true && table[4][i].draw == true) {
					score := 0
					for _, row := range table {
						for _, column := range row {
							if column.draw == false {
								score += column.number
							}
						}
					}
					score = score * num
					fmt.Println("Ans :", score)
					return
				}
			}
		}
	}
}

func day04_II(tables [][][]cell, nums []int) {
	for _, num := range nums {
		for i, table := range tables {
			for j, row := range table {
				for k, column := range row {
					if column.number == num {
						tables[i][j][k].draw = true
					}
				}
			}
		}

		if len(tables) != 1 {
			for _, table := range tables {
				for i := 0; i < 5; i++ {
					if (table[i][0].draw == true && table[i][1].draw == true && table[i][2].draw == true && table[i][3].draw == true && table[i][4].draw == true) || (table[0][i].draw == true && table[1][i].draw == true && table[2][i].draw == true && table[3][i].draw == true && table[4][i].draw == true) {

						tables = remove(tables, table)
						break
					}
				}
			}
		} else {
			table := tables[0]
			for i := 0; i < 5; i++ {
				if (table[i][0].draw == true && table[i][1].draw == true && table[i][2].draw == true && table[i][3].draw == true && table[i][4].draw == true) || (table[0][i].draw == true && table[1][i].draw == true && table[2][i].draw == true && table[3][i].draw == true && table[4][i].draw == true) {
					score := 0
					for _, row := range tables[0] {
						for _, column := range row {
							if column.draw == false {
								score += column.number
							}
						}
					}
					score = score * num
					fmt.Println(table)
					fmt.Println("Ans :", score)
					return
				}
			}
		}
	}
}

func main() {
	read_input()
	nums, tables := read_input()

	day04_II(tables, nums)
}

File: Python/day04.py

file = open("inputs/day04.txt", "r")

nums = file.readline()
nums = nums.replace("\n", "").split(",")

file.readline()
input = file.read()
input = input.split("\n\n")
tables=[]

def day04_I(input: list):
    print("====== Part I ======")
    for table in input:
        table = table.split("\n")
        new_table = []
        for row in table:
            row = row.replace("  ", " ").split(" ")
            if row[0] == '':
                row.pop(0)
            new_row = []
            for column in row:
                new_row.append([int(column), 0])
            new_table.append(new_row)
        tables.append(new_table)

    for num in nums:
        for i, table in enumerate(tables):
            for j, row in enumerate(table):
                for k, column in enumerate(row):
                    if column[0] == int(num):
                        tables[i][j][k][1] = 1
        
        for table_number, table in enumerate(tables):
            for i in range (5):
                if (table[i][0][1] == 1 and table[i][1][1] == 1 and table[i][2][1] == 1 and table[i][3][1] == 1 and table[i][4][1] == 1) or (table[0][i][1] and table[1][i][1] and table[2][i][1] and table[3][i][1] and table[4][i][1]):
                    print("table number " + str(num) + " won")
                    score = 0
                    for row in table:
                        for column in row:
                            if column[1] == 0:
                                score += column[0]
                    score = score * int(num)
                    print("Ans : " + str(score))
                    exit()

def day04_II(input : list):
    print("====== Part II ======")
    for table in input:
        table = table.split("\n")
        new_table = []
        for row in table:
            row = row.replace("  ", " ").split(" ")
            if row[0] == '':
                row.pop(0)
            new_row = []
            for column in row:
                new_row.append([int(column), 0])
            new_table.append(new_row)
        tables.append(new_table)

    for num in nums:
        for i, table in enumerate(tables):
            for j, row in enumerate(table):
                for k, column in enumerate(row):
                    if column[0] == int(num):
                        tables[i][j][k][1] = 1
        
        if len(tables) != 1:
            for j, table in enumerate(tables):
                for i in range (5):
                    if (table[i][0][1] == 1 and table[i][1][1] == 1 and table[i][2][1] == 1 and table[i][3][1] == 1 and table[i][4][1] == 1) or (table[0][i][1] == 1 and table[1][i][1] == 1 and table[2][i][1] == 1 and table[3][i][1] == 1 and table[4][i][1] == 1 ):
                        tables.remove(table)
                        break
        else:
            table = tables[0]
            for i in range(5):
                if (table[i][0][1] == 1 and table[i][1][1] == 1 and table[i][2][1] == 1 and table[i][3][1] == 1 and table[i][4][1] == 1) or (table[0][i][1] == 1 and table[1][i][1] == 1 and table[2][i][1] == 1 and table[3][i][1] == 1 and table[4][i][1] == 1 ):
                    score = 0
                    for row in tables[0]:
                        for column in row:
                            if column[1] == 0:
                                score += column[0]
                    score = score * int(num)
                    print("Ans : " + str(score))
                    exit()

day04_II(input)
